File: src/simulation/economy.js
import { recordEvent } from "./journal.js";
import { technologyEfficiency } from "./technology.js";
import { dailyHouseholdNeed } from "./needs.js";

const EPSILON = 1e-9;

function quantize(value) {
  return Math.round(value * 1_000_000) / 1_000_000;
}

function seasonalityFactor(world, recipe) {
  if (!recipe.seasonality) return 1;
  const dayOfYear = world.day % world.calendar.daysPerYear;
  const wave = 0.5 + 0.5 * Math.cos(
    Math.PI * 2 * (dayOfYear - recipe.seasonality.peakDay) / world.calendar.daysPerYear
  );
  return recipe.seasonality.minimum + (1 - recipe.seasonality.minimum) * wave;
}

function siteFactor(world, industry, recipe) {
  if (!recipe.sitePotential) return 1;
  const territory = world.spatial.territories[industry.zoneId];
  const basePotential = territory.resourcePotential[recipe.sitePotential];
  let remainingRatio = 1;
  if (recipe.sitePotential === "arable" || recipe.sitePotential === "pasture") {
    remainingRatio = territory.fertility > 0
      ? Math.max(0.15, territory.naturalState.soilQuality / territory.fertility)
      : 0.15;
  } else if (recipe.sitePotential === "timber") {
    remainingRatio = territory.forestCover > 0
      ? Math.max(0.05, territory.naturalState.forestBiomass / territory.forestCover)
      : 0;
  } else if (recipe.sitePotential === "fish") {
    remainingRatio = basePotential > 0
      ? Math.max(0.08, territory.naturalState.fishStock / basePotential)
      : 0;
  } else if (["clay", "stone", "iron_ore"].includes(recipe.sitePotential)) {
    remainingRatio = territory.naturalState.deposits[recipe.sitePotential];
  }
  return 0.18 + basePotential * remainingRatio * 0.82;
}

function recoverNaturalSites(world) {
  const siteIds = new Set(Object.values(world.cities)
    .flatMap((city) => city.industries.map((industry) => industry.zoneId)));
  for (const territoryId of [...siteIds].sort()) {
    const territory = world.spatial.territories[territoryId];
    const natural = territory.naturalState;
    natural.soilQuality = quantize(natural.soilQuality +
      (territory.fertility - natural.soilQuality) * 0.00045);
    natural.forestBiomass = quantize(natural.forestBiomass +
      (territory.forestCover - natural.forestBiomass) * 0.0003);
    natural.fishStock = quantize(natural.fishStock +
      (territory.resourcePotential.fish - natural.fishStock) * 0.004);
  }
}

function applyEnvironmentalUse(city, industry, recipe, territory, batches) {
  if (!recipe.sitePotential || batches <= 0) return;
  const natural = territory.naturalState;
  const potential = recipe.sitePotential;
  if (potential === "arable") {
    const rotation = city.technologyState.crop_rotation?.adoption ?? 0;
    natural.soilQuality = quantize(Math.max(0.05,
      natural.soilQuality - batches * 0.000055 * (1 - rotation * 0.65)
    ));
  } else if (potential === "pasture") {
    natural.soilQuality = quantize(Math.max(0.05, natural.soilQuality - batches * 0.000025));
  } else if (potential === "timber") {
    natural.forestBiomass = quantize(Math.max(0,
      natural.forestBiomass - batches * 0.00014
    ));
  } else if (potential === "fish") {
    natural.fishStock = quantize(Math.max(0, natural.fishStock - batches * 0.00016));
  } else if (["clay", "stone", "iron_ore"].includes(potential)) {
    const depletion = { clay: 0.000025, stone: 0.000018, iron_ore: 0.00007 }[potential];
    natural.deposits[potential] = quantize(Math.max(0,
      natural.deposits[potential] - batches * depletion
    ));
  }
  natural.extractedBatches[potential] = quantize(
    (natural.extractedBatches[potential] ?? 0) + batches
  );
}

function decayStocks(world, content, telemetry) {
  const resourceById = new Map(content.resources.resources.map((resource) => [resource.id, resource]));
  for (const city of Object.values(world.cities)) {
    for (const resourceId of Object.keys(city.stocks).sort()) {
      const decay = resourceById.get(resourceId).decayPerDay;
      if (decay <= 0 || city.stocks[resourceId] <= 0) continue;
      const lost = quantize(city.stocks[resourceId] * decay);
      city.stocks[resourceId] = quantize(Math.max(0, city.stocks[resourceId] - lost));
      telemetry.decayedByResource[resourceId] = quantize(
        (telemetry.decayedByResource[resourceId] ?? 0) + lost
      );
    }
  }
}

function updateConstraint(world, industry, recipe, city, plannedBatches, actualBatches, constraints, causeIds) {
  const key = constraints.join("|");
  if (constraints.length === 0) {
    if (industry.lastConstraintKey !== null) {
      recordEvent(world, {
        type: "production_restored",
        subjectId: industry.id,
        causeIds: industry.constraintEventId ? [industry.constraintEventId] : [],
        details: { cityId: city.id, recipeId: recipe.id }
      });
    }
    industry.lastConstraintKey = null;
    industry.constraintEventId = null;
    return;
  }
  if (key === industry.lastConstraintKey) return;
  const event = recordEvent(world, {
    type: "production_constrained",
    subjectId: industry.id,
    causeIds,
    details: { cityId: city.id, recipeId: recipe.id, plannedBatches, actualBatches, constraints }
  });
  industry.lastConstraintKey = key;
  industry.constraintEventId = event.id;
  for (const resourceId of Object.keys(recipe.outputs).sort()) city.resourceSignals[resourceId] = event.id;
}

function produce(world, content, telemetry) {
  const recipeById = new Map(content.recipes.recipes.map((recipe) => [recipe.id, recipe]));
  for (const cityId of Object.keys(world.cities).sort()) {
    const city = world.cities[cityId];
    const population = world.spatial.nodes[city.spatialNodeId].aggregate.population;
    const activeEffects = Object.entries(city.activeEffects).sort(([left], [right]) => left.localeCompare(right));
    const workforceMultiplier = activeEffects.reduce((value, [, effect]) => value * effect.multiplier, 1);
    let availableLabor = population * city.workerShare * workforceMultiplier;

    for (const industry of city.industries) {
      const recipe = recipeById.get(industry.recipeId);
      const plannedBatches = quantize(industry.capacity * seasonalityFactor(world, recipe) *
        siteFactor(world, industry, recipe) * technologyEfficiency(city, recipe));
      const laborLimit = availableLabor / recipe.laborPerBatch;
      let inputLimit = Infinity;
      for (const [resourceId, amount] of Object.entries(recipe.inputs).sort()) {
        if (amount > 0) inputLimit = Math.min(inputLimit, city.stocks[resourceId] / amount);
      }
      const batches = quantize(Math.max(0, Math.min(plannedBatches, laborLimit, inputLimit)));

      for (const [resourceId, amount] of Object.entries(recipe.inputs).sort()) {
        const consumed = quantize(amount * batches);
        city.stocks[resourceId] = quantize(Math.max(0, city.stocks[resourceId] - consumed));
        telemetry.industrialConsumptionByResource[resourceId] = quantize(
          (telemetry.industrialConsumptionByResource[resourceId] ?? 0) + consumed
        );
      }
      for (const [resourceId, amount] of Object.entries(recipe.outputs).sort()) {
        const produced = quantize(amount * batches);
        city.stocks[resourceId] = quantize(city.stocks[resourceId] + produced);
        telemetry.productionByResource[resourceId] = quantize(
          (telemetry.productionByResource[resourceId] ?? 0) + produced
        );
      }
      availableLabor = quantize(Math.max(0, availableLabor - recipe.laborPerBatch * batches));
      industry.totalBatches = quantize(industry.totalBatches + batches);
      applyEnvironmentalUse(city, industry, recipe, world.spatial.territories[industry.zoneId], batches);

      const constraints = [];
      const causeIds = [];
      if (laborLimit + EPSILON < plannedBatches) {
        constraints.push("labor");
        causeIds.push(...activeEffects.map(([, effect]) => effect.startEventId));
      }
      for (const [resourceId, amount] of Object.entries(recipe.inputs).sort()) {
        if (amount > 0 && city.stocks[resourceId] + amount * batches < amount * plannedBatches - EPSILON) {
          constraints.push(`input:${resourceId}`);
          if (city.resourceSignals[resourceId]) causeIds.push(city.resourceSignals[resourceId]);
        }
      }
      updateConstraint(world, industry, recipe, city, plannedBatches, batches, constraints, causeIds);
    }
  }
}

function consumeFood(world, telemetry) {
  for (const cityId of Object.keys(world.cities).sort()) {
    const city = world.cities[cityId];
    const population = world.spatial.nodes[city.spatialNodeId].aggregate.population;
    const needed = quantize(population * city.foodPerPersonPerDay);
    const consumed = quantize(Math.min(needed, city.stocks.food));
    const missing = quantize(needed - consumed);
    city.stocks.food = quantize(Math.max(0, city.stocks.food - consumed));
    telemetry.householdFoodConsumed = quantize(telemetry.householdFoodConsumed + consumed);
    telemetry.householdFoodMissing = quantize(telemetry.householdFoodMissing + missing);
    telemetry.householdConsumptionByResource.food = quantize(
      (telemetry.householdConsumptionByResource.food ?? 0) + consumed
    );
    telemetry.householdMissingByResource.food = quantize(
      (telemetry.householdMissingByResource.food ?? 0) + missing
    );

    if (missing > EPSILON) {
      city.shortage.days += 1;
      city.shortage.episodeDays += 1;
      city.shortage.missingStreak += 1;
      city.shortage.satisfiedStreak = 0;
      city.shortage.totalFoodMissing = quantize(city.shortage.totalFoodMissing + missing);
      if (!city.shortage.active && city.shortage.missingStreak >= 2) {
        const event = recordEvent(world, {
          type: "food_shortage_started",
          subjectId: cityId,
          causeIds: city.resourceSignals.food ? [city.resourceSignals.food] : [],
          details: { cityId, needed, available: consumed, missing }
        });
        city.shortage.active = true;
        city.shortage.eventId = event.id;
      }
    } else {
      city.shortage.missingStreak = 0;
      if (!city.shortage.active) city.shortage.episodeDays = 0;
      if (city.shortage.active) {
        city.shortage.satisfiedStreak += 1;
        if (city.shortage.satisfiedStreak >= 3) {
          recordEvent(world, {
            type: "food_shortage_ended",
            subjectId: cityId,
            causeIds: [city.shortage.eventId],
            details: { cityId, durationDays: city.shortage.episodeDays }
          });
          city.shortage.active = false;
          city.shortage.episodeDays = 0;
          city.shortage.satisfiedStreak = 0;
          city.shortage.eventId = null;
        }
      }
    }
  }
}

function consumeOtherHouseholdNeeds(world, content, telemetry) {
  for (const cityId of Object.keys(world.cities).sort()) {
    const city = world.cities[cityId];
    for (const resource of content.resources.resources.filter((item) =>
      item.householdNeed && item.id !== "food"
    ).sort((left, right) => left.id.localeCompare(right.id))) {
      const state = city.needs[resource.id];
      const needed = quantize(dailyHouseholdNeed(world, city, resource));
      const consumed = quantize(Math.min(needed, city.stocks[resource.id]));
      const missing = quantize(needed - consumed);
      city.stocks[resource.id] = quantize(Math.max(0, city.stocks[resource.id] - consumed));
      telemetry.householdConsumptionByResource[resource.id] = quantize(
        (telemetry.householdConsumptionByResource[resource.id] ?? 0) + consumed
      );
      telemetry.householdMissingByResource[resource.id] = quantize(
        (telemetry.householdMissingByResource[resource.id] ?? 0) + missing
      );
      if (missing > EPSILON) {
        state.days += 1;
        state.episodeDays += 1;
        state.missingStreak += 1;
        state.satisfiedStreak = 0;
        state.totalMissing = quantize(state.totalMissing + missing);
        if (!state.active && state.missingStreak >= 3) {
          const event = recordEvent(world, {
            type: "resource_shortage_started",
            subjectId: `${cityId}:${resource.id}`,
            causeIds: city.resourceSignals[resource.id] ? [city.resourceSignals[resource.id]] : [],
            details: { cityId, resourceId: resource.id, needed, available: consumed, missing }
          });
          state.active = true;
          state.eventId = event.id;
        }
      } else {
        state.missingStreak = 0;
        if (!state.active) state.episodeDays = 0;
        if (state.active) {
          state.satisfiedStreak += 1;
          if (state.satisfiedStreak >= 3) {
            recordEvent(world, {
              type: "resource_shortage_ended",
              subjectId: `${cityId}:${resource.id}`,
              causeIds: state.eventId ? [state.eventId] : [],
              details: { cityId, resourceId: resource.id, durationDays: state.episodeDays }
            });
            state.active = false;
            state.episodeDays = 0;
            state.satisfiedStreak = 0;
            state.eventId = null;
          }
        }
      }
    }
  }
}

export function runEconomyDay(world, content) {
  const telemetry = {
    day: world.day,
    productionByResource: {},
    industrialConsumptionByResource: {},
    decayedByResource: {},
    householdFoodConsumed: 0,
    householdFoodMissing: 0,
    householdConsumptionByResource: {},
    householdMissingByResource: {},
    infrastructureConsumptionByResource: {},
    shipmentsDispatched: 0,
    shipmentsArrived: 0
  };
  decayStocks(world, content, telemetry);
  recoverNaturalSites(world);
  produce(world, content, telemetry);
  consumeFood(world, telemetry);
  consumeOtherHouseholdNeeds(world, content, telemetry);
  return telemetry;
}

File: src/WorldGen.Core/Simulation/Economy.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using WorldGen.Core.Content;

namespace WorldGen.Core.Simulation;

public static class Economy
{
    private const double Epsilon = 1e-9;

    public static DailyTelemetry RunDay(WorldState world, ContentCatalog content, IReadOnlyDictionary<string, double>? utilization = null, SettlementSimulation? development = null)
    {
        var telemetry = new DailyTelemetry { Day = world.Day };
        development?.ReconcileFoodComposition();
        DecayStocks(world, content, telemetry);
        development?.ReconcileFoodComposition();
        if (development is null) RecoverNaturalSites(world); else development.RecoverNaturalSites();
        development?.RunDay(telemetry);
        Produce(world, content, telemetry, utilization, development);
        ConsumeFood(world, telemetry, development);
        ConsumeOtherNeeds(world, content, telemetry, development);
        development?.EvaluateWellbeing();
        return telemetry;
    }

    private static void DecayStocks(WorldState world, ContentCatalog content, DailyTelemetry telemetry)
    {
        var resources = content.Resources.Resources.ToDictionary(resource => resource.Id, StringComparer.Ordinal);
        foreach (var city in world.Cities.Values)
        foreach (var resourceId in city.Stocks.Keys.Order(StringComparer.Ordinal).ToArray())
        {
            var decay = resources[resourceId].DecayPerDay;
            if (decay <= 0 || city.Stocks[resourceId] <= 0) continue;
            var lost = SimulationMath.Quantize(city.Stocks[resourceId] * decay);
            city.Stocks[resourceId] = SimulationMath.Quantize(Math.Max(0, city.Stocks[resourceId] - lost));
            telemetry.DecayedByResource[resourceId] = SimulationMath.Quantize(telemetry.DecayedByResource.GetValueOrDefault(resourceId) + lost);
        }
    }

    private static void RecoverNaturalSites(WorldState world)
    {
        var ids = world.Cities.Values.SelectMany(city => city.Industries.Select(industry => industry.ZoneId))
            .ToHashSet(StringComparer.Ordinal);
        foreach (var id in ids.Order(StringComparer.Ordinal))
        {
            var territory = world.Spatial.Territories[id];
            var natural = territory.NaturalState;
            natural.SoilQuality = SimulationMath.Quantize(natural.SoilQuality + (territory.Fertility - natural.SoilQuality) * 0.00045);
            natural.ForestBiomass = SimulationMath.Quantize(natural.ForestBiomass + (territory.ForestCover - natural.ForestBiomass) * 0.0003);
            natural.FishStock = SimulationMath.Quantize(natural.FishStock + (territory.ResourcePotential["fish"] - natural.FishStock) * 0.004);
        }
    }

    private static double TechnologyEfficiency(CityState city, RecipeDefinition recipe)
    {
        if (recipe.RequiredTechnologyIds.Count == 0) return 1;
        var practice = recipe.RequiredTechnologyIds.Select(id => city.TechnologyState[id])
            .Min(state => Math.Min(Math.Min(state.Knowledge, state.Competence), Math.Min(state.Capability, state.Adoption)));
        return 0.22 + practice * 0.78;
    }

    private static double SeasonalityFactor(WorldState world, RecipeDefinition recipe) =>
        NeedsAndDemand.SeasonalNeedFactor(world, recipe.Seasonality);

    private static double SiteFactor(WorldState world, IndustryState industry, RecipeDefinition recipe)
    {
        if (recipe.SitePotential is null) return 1;
        var territory = world.Spatial.Territories[industry.ZoneId];
        var potential = recipe.SitePotential;
        var basePotential = territory.ResourcePotential[potential];
        var ratio = 1d;
        if (potential is "arable" or "pasture") ratio = territory.Fertility > 0 ? Math.Max(0.15, territory.NaturalState.SoilQuality / territory.Fertility) : 0.15;
        else if (potential == "timber") ratio = territory.ForestCover > 0 ? Math.Max(0.05, territory.NaturalState.ForestBiomass / territory.ForestCover) : 0;
        else if (potential == "fish") ratio = basePotential > 0 ? Math.Max(0.08, territory.NaturalState.FishStock / basePotential) : 0;
        else if (potential is "clay" or "stone" or "iron_ore") ratio = territory.NaturalState.Deposits[potential];
        return 0.18 + basePotential * ratio * 0.82;
    }

    private static void Produce(WorldState world, ContentCatalog content, DailyTelemetry telemetry, IReadOnlyDictionary<string, double>? utilization, SettlementSimulation? development)
    {
        var recipes = content.Recipes.Recipes.ToDictionary(recipe => recipe.Id, StringComparer.Ordinal);
        foreach (var cityId in world.Cities.Keys.Order(StringComparer.Ordinal))
        {
            var city = world.Cities[cityId];
            var population = world.Spatial.Nodes[city.SpatialNodeId].Aggregate.Population;
            var effects = city.ActiveEffects.OrderBy(pair => pair.Key, StringComparer.Ordinal).ToArray();
            var workforceMultiplier = effects.Aggregate(1d, (value, pair) => value * pair.Value.Multiplier);
            var availableLabor = Math.Max(0, population * city.WorkerShare * workforceMultiplier - (development?.ReservedWorkerDays(cityId) ?? 0));
            foreach (var industry in city.Industries)
            {
                var recipe = recipes[industry.RecipeId];
                var planned = SimulationMath.Quantize(industry.Capacity * SeasonalityFactor(world, recipe) *
                    SiteFactor(world, industry, recipe) * TechnologyEfficiency(city, recipe) *
                    (utilization?.GetValueOrDefault(industry.Id, 1) ?? 1));
                var laborLimit = availableLabor / recipe.LaborPerBatch;
                var inputLimit = double.PositiveInfinity;
                foreach (var pair in recipe.Inputs.OrderBy(pair => pair.Key, StringComparer.Ordinal))
                    if (pair.Value > 0) inputLimit = Math.Min(inputLimit, city.Stocks[pair.Key] / pair.Value);
                var batches = SimulationMath.Quantize(Math.Max(0, Math.Min(planned, Math.Min(laborLimit, inputLimit))));
                if (development is not null) batches = Math.Min(batches, Math.Min(laborLimit, inputLimit));
                batches = development?.LimitIndustry(world.Spatial.Territories[industry.ZoneId], recipe, batches) ?? batches;
                foreach (var pair in recipe.Inputs.OrderBy(pair => pair.Key, StringComparer.Ordinal))
                {
                    var consumed = SimulationMath.Quantize(pair.Value * batches);
                    city.Stocks[pair.Key] = SimulationMath.Quantize(Math.Max(0, city.Stocks[pair.Key] - consumed));
                    telemetry.IndustrialConsumptionByResource[pair.Key] = SimulationMath.Quantize(telemetry.IndustrialConsumptionByResource.GetValueOrDefault(pair.Key) + consumed);
                }
                foreach (var pair in recipe.Outputs.OrderBy(pair => pair.Key, StringComparer.Ordinal))
                {
                    var produced = SimulationMath.Quantize(pair.Value * batches);
                    city.Stocks[pair.Key] = SimulationMath.Quantize(city.Stocks[pair.Key] + produced);
                    telemetry.ProductionByResource[pair.Key] = SimulationMath.Quantize(telemetry.ProductionByResource.GetValueOrDefault(pair.Key) + produced);
                }
                availableLabor = SimulationMath.Quantize(Math.Max(0, availableLabor - recipe.LaborPerBatch * batches));
                development?.RecordIndustryLabor(cityId, recipe.LaborPerBatch * batches);
                industry.TotalBatches = SimulationMath.Quantize(industry.TotalBatches + batches);
                if (development?.UseIndustryResource(world.Spatial.Territories[industry.ZoneId], recipe, batches) != true)
                    ApplyEnvironmentalUse(city, industry, recipe, world.Spatial.Territories[industry.ZoneId], batches);

                var constraints = new List<string>();
                var causes = new List<string?>();
                if (laborLimit + Epsilon < planned) { constraints.Add("labor"); causes.AddRange(effects.Select(pair => pair.Value.StartEventId)); }
                foreach (var pair in recipe.Inputs.OrderBy(pair => pair.Key, StringComparer.Ordinal))
                    if (pair.Value > 0 && city.Stocks[pair.Key] + pair.Value * batches < pair.Value * planned - Epsilon)
                    { constraints.Add($"input:{pair.Key}"); causes.Add(city.ResourceSignals.GetValueOrDefault(pair.Key)); }
                UpdateConstraint(world, industry, recipe, city, planned, batches, constraints, causes);
            }
        }
    }

    private static void ApplyEnvironmentalUse(CityState city, IndustryState industry, RecipeDefinition recipe,
        Spatial.Territory territory, double batches)
    {
        if (recipe.SitePotential is null || batches <= 0) return;
        var natural = territory.NaturalState;
        var potential = recipe.SitePotential;
        if (potential == "arable") natural.SoilQuality = SimulationMath.Quantize(Math.Max(0.05, natural.SoilQuality - batches * 0.000055 * (1 - city.TechnologyState.GetValueOrDefault("crop_rotation")?.Adoption * 0.65 ?? 1)));
        else if (potential == "pasture") natural.SoilQuality = SimulationMath.Quantize(Math.Max(0.05, natural.SoilQuality - batches * 0.000025));
        else if (potential == "timber") natural.ForestBiomass = SimulationMath.Quantize(Math.Max(0, natural.ForestBiomass - batches * 0.00014));
        else if (potential == "fish") natural.FishStock = SimulationMath.Quantize(Math.Max(0, natural.FishStock - batches * 0.00016));
        else if (potential is "clay" or "stone" or "iron_ore")
        {
            var depletion = potential == "clay" ? 0.000025 : potential == "stone" ? 0.000018 : 0.00007;
            natural.Deposits[potential] = SimulationMath.Quantize(Math.Max(0, natural.Deposits[potential] - batches * depletion));
        }
        natural.ExtractedBatches[potential] = SimulationMath.Quantize(natural.ExtractedBatches.GetValueOrDefault(potential) + batches);
    }

    private static void UpdateConstraint(WorldState world, IndustryState industry, RecipeDefinition recipe, CityState city,
        double planned, double actual, List<string> constraints, List<string?> causes)
    {
        var key = string.Join('|', constraints);
        if (constraints.Count == 0)
        {
            if (industry.LastConstraintKey is not null)
                Journal.Record(world, "production_restored", industry.Id,
                    industry.ConstraintEventId is null ? [] : [industry.ConstraintEventId],
                    new JsonObject { ["cityId"] = city.Id, ["recipeId"] = recipe.Id });
            industry.LastConstraintKey = null; industry.ConstraintEventId = null; return;
        }
        if (key == industry.LastConstraintKey) return;
        var evt = Journal.Record(world, "production_constrained", industry.Id, causes,
            new JsonObject { ["cityId"] = city.Id, ["recipeId"] = recipe.Id, ["plannedBatches"] = planned,
                ["actualBatches"] = actual, ["constraints"] = JsonSerializer.SerializeToNode(constraints) });
        industry.LastConstraintKey = key; industry.ConstraintEventId = evt.Id;
        foreach (var resourceId in recipe.Outputs.Keys.Order(StringComparer.Ordinal)) city.ResourceSignals[resourceId] = evt.Id;
    }

    private static void ConsumeFood(WorldState world, DailyTelemetry telemetry, SettlementSimulation? development)
    {
        foreach (var cityId in world.Cities.Keys.Order(StringComparer.Ordinal))
        {
            var city = world.Cities[cityId];
            var population = world.Spatial.Nodes[city.SpatialNodeId].Aggregate.Population;
            var needed = SimulationMath.Quantize(population * city.FoodPerPersonPerDay);
            var localNeed = Math.Max(0, needed - (development?.RemoteNeed(cityId, "food") ?? 0));
            var localConsumed = SimulationMath.Quantize(Math.Min(localNeed, city.Stocks["food"]));
            development?.RecordFoodConsumption(city, localConsumed);
            var consumed = SimulationMath.Quantize(localConsumed + (development?.RemoteConsumption(cityId, "food") ?? 0));
            var missing = SimulationMath.Quantize(needed - consumed);
            city.Stocks["food"] = SimulationMath.Quantize(Math.Max(0, city.Stocks["food"] - localConsumed));
            telemetry.HouseholdFoodConsumed = SimulationMath.Quantize(telemetry.HouseholdFoodConsumed + consumed);
            telemetry.HouseholdFoodMissing = SimulationMath.Quantize(telemetry.HouseholdFoodMissing + missing);
            telemetry.HouseholdConsumptionByResource["food"] = SimulationMath.Quantize(telemetry.HouseholdConsumptionByResource.GetValueOrDefault("food") + consumed);
            telemetry.HouseholdMissingByResource["food"] = SimulationMath.Quantize(telemetry.HouseholdMissingByResource.GetValueOrDefault("food") + missing);
            if (missing > Epsilon)
            {
                city.Shortage.Days++; city.Shortage.EpisodeDays++; city.Shortage.MissingStreak++; city.Shortage.SatisfiedStreak = 0;
                city.Shortage.TotalFoodMissing = SimulationMath.Quantize(city.Shortage.TotalFoodMissing + missing);
                if (!city.Shortage.Active && city.Shortage.MissingStreak >= 2)
                {
                    var evt = Journal.Record(world, "food_shortage_started", cityId,
                        city.ResourceSignals.TryGetValue("food", out var cause) ? [cause] : [],
                        new JsonObject { ["cityId"] = cityId, ["needed"] = needed, ["available"] = consumed, ["missing"] = missing });
                    city.Shortage.Active = true; city.Shortage.EventId = evt.Id;
                }
            }
            else
            {
                city.Shortage.MissingStreak = 0;
                if (!city.Shortage.Active) city.Shortage.EpisodeDays = 0;
                if (city.Shortage.Active && ++city.Shortage.SatisfiedStreak >= 3)
                {
                    Journal.Record(world, "food_shortage_ended", cityId, [city.Shortage.EventId],
                        new JsonObject { ["cityId"] = cityId, ["durationDays"] = city.Shortage.EpisodeDays });
                    city.Shortage.Active = false; city.Shortage.EpisodeDays = 0; city.Shortage.SatisfiedStreak = 0; city.Shortage.EventId = null;
                }
            }
        }
    }

    private static void ConsumeOtherNeeds(WorldState world, ContentCatalog content, DailyTelemetry telemetry, SettlementSimulation? development)
    {
        var resources = content.Resources.Resources.Where(resource => resource.HouseholdNeed is not null && resource.Id != "food")
            .OrderBy(resource => resource.Id, StringComparer.Ordinal);
        foreach (var cityId in world.Cities.Keys.Order(StringComparer.Ordinal))
        foreach (var resource in resources)
        {
            var city = world.Cities[cityId]; var state = city.Needs[resource.Id];
            var needed = SimulationMath.Quantize(NeedsAndDemand.DailyHouseholdNeed(world, city, resource));
            var localNeed = Math.Max(0, needed - (development?.RemoteNeed(cityId, resource.Id) ?? 0));
            var localConsumed = SimulationMath.Quantize(Math.Min(localNeed, city.Stocks[resource.Id]));
            var consumed = SimulationMath.Quantize(localConsumed + (development?.RemoteConsumption(cityId, resource.Id) ?? 0));
            var missing = SimulationMath.Quantize(needed - consumed);
            city.Stocks[resource.Id] = SimulationMath.Quantize(Math.Max(0, city.Stocks[resource.Id] - localConsumed));
            telemetry.HouseholdConsumptionByResource[resource.Id] = SimulationMath.Quantize(telemetry.HouseholdConsumptionByResource.GetValueOrDefault(resource.Id) + consumed);
            telemetry.HouseholdMissingByResource[resource.Id] = SimulationMath.Quantize(telemetry.HouseholdMissingByResource.GetValueOrDefault(resource.Id) + missing);
            if (missing > Epsilon)
            {
                state.Days++; state.EpisodeDays++; state.MissingStreak++; state.SatisfiedStreak = 0;
                state.TotalMissing = SimulationMath.Quantize(state.TotalMissing + missing);
                if (!state.Active && state.MissingStreak >= 3)
                {
                    var evt = Journal.Record(world, "resource_shortage_started", $"{cityId}:{resource.Id}",
                        city.ResourceSignals.TryGetValue(resource.Id, out var cause) ? [cause] : [],
                        new JsonObject { ["cityId"] = cityId, ["resourceId"] = resource.Id, ["needed"] = needed, ["available"] = consumed, ["missing"] = missing });
                    state.Active = true; state.EventId = evt.Id;
                }
            }
            else
            {
                state.MissingStreak = 0; if (!state.Active) state.EpisodeDays = 0;
                if (state.Active && ++state.SatisfiedStreak >= 3)
                {
                    Journal.Record(world, "resource_shortage_ended", $"{cityId}:{resource.Id}", [state.EventId],
                        new JsonObject { ["cityId"] = cityId, ["resourceId"] = resource.Id, ["durationDays"] = state.EpisodeDays });
                    state.Active = false; state.EpisodeDays = 0; state.SatisfiedStreak = 0; state.EventId = null;
                }
            }
        }
    }
}
